packages/enzyme/src/selectors.js

```diff
@@ -0,0 +1,318 @@
+ */
+function uniqueReduce(fn, nodes) {
+  return unique(nodes.reduce(fn, []));
+}
+
+/**
+ * Takes a CSS selector and returns a set of tokens parsed
+ * by scalpel.
+ * @param {String} selector 
+ */
+function safelyGenerateTokens(selector) {
+  try {
+    return parser.parse(selector);
+  } catch (err) {
+    throw new Error(`Failed to parse selector: ${selector}`);
+  }
+}
+
+/**
+ * Takes a node and a token and determines if the node
+ * matches the predicate defined by the token.
+ * @param {Node} node 
+ * @param {Token} token 
+ */
+function nodeMatchesToken(node, token) {
+  if (node === null || typeof node === 'string') {
+    return false;
+  }
+  switch (token.type) {
+    /**
+     * Match against the className prop
+     * @example '.active' matches <div className='active' />
+     */
+    case CLASS_SELECTOR:
+      return hasClassName(node, token.name);
+    /**
+     * Simple type matching
+     * @example 'div' matches <div />
+     */
+    case TYPE_SELECTOR:
+      return nodeHasType(node, token.name);
+    /**
+     * Match against the `id` prop
+     * @example '#nav' matches <ul id="nav" />
+     */
+    case ID_SELECTOR:
+      return nodeHasId(node, token.name);
+    /**
+     * Matches if an attribute is present, regardless
+     * of its value
+     * @example '[disabled]' matches <a disabled />
+     */
+    case ATTRIBUTE_PRESENCE:
+      return nodeHasProperty(node, token.name);
+    /**
+     * Matches if an attribute is present with the
+     * provided value
+     * @example '[data-foo=foo]' matches <div data-foo="foo" />
+     */
+    case ATTRIBUTE_VALUE:
+      return nodeHasProperty(node, token.name, token.value);
+    case PSEUDO_ELEMENT:
+    case PSEUDO_CLASS:
+      throw new Error('Enzyme::Selector does not support psuedo-element or psuedo-class selectors.');
+    default:
+      throw new Error(`Unknown token type: ${token.type}`);
+  }
+}
+
+/**
+ * Returns a predicate function that checks if a
+ * node matches every token in the body of a selector
+ * token.
+ * @param {Token} token 
+ */
+function buildPredicateFromToken(token) {
+  return node => token.body.every(
+    bodyToken => nodeMatchesToken(node, bodyToken),
+  );
+}
+
+/**
+ * Returns whether a parsed selector is a complex selector, which
+ * is defined as a selector that contains combinators.
+ * @param {Array<Token>} tokens 
+ */
+function isComplexSelector(tokens) {
+  return tokens.some(token => token.type !== SELECTOR);
+}
+
+
+/**
+ * Takes a component constructor, object, or string representing
+ * a simple selector and returns a predicate function that can
+ * be applied to a single node. 
+ * @param {Function|Object|String} selector 
+ */
+export function buildPredicate(selector) {
+  switch (typeof selector) {
+    case 'function':
+      // constructor
+      return node => node && node.type === selector;
+    case 'object':
+      if (!Array.isArray(selector) && selector !== null && !isEmpty(selector)) {
```

Comment:
what does lodash return for `isEmpty(null)`?

if we're keeping `!== null` for perf, then we should be checking that first, before `isArray`.

Reply:
I'll figure out what it returns and update this. I just copied what was already there FWIW: https://github.com/airbnb/enzyme/blob/c15dd479684a9e3423d3efb5aea592fcfde1a842/packages/enzyme/src/RSTTraversal.js#L119